src/main/java/org/spongepowered/api/service/metrics/pullgauge/PullGauge.java

```diff
@@ -0,0 +1,168 @@
+
+import io.prometheus.client.Collector.MetricFamilySamples;
+import io.prometheus.client.Collector.MetricFamilySamples.Sample;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.ToDoubleFunction;
+
+import static io.prometheus.client.Collector.Type.GAUGE;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+
+public abstract class PullGauge<T> {
+    private final String name;
+    private final String unit;
+    private final String help;
+
+    public PullGauge(String name, String unit, String help) {
+        this.name = name;
+        this.unit = unit;
+        this.help = help;
+    }
+
+    protected final Sample sample(List<String> labelNames, List<String> labelValues, double value) {
+        return new Sample(name, labelNames, labelValues, value);
+    }
+
+    protected final Sample sample(double value) {
+        return sample(emptyList(), emptyList(), value);
+    }
+
+    protected final MetricFamilySamples bundle(List<Sample> samples) {
+        return new MetricFamilySamples(name, unit, GAUGE, help, samples);
+    }
+
+    protected abstract List<Sample> makeSamples(T subject);
+
+    public final MetricFamilySamples sample(T subject) {
+        return bundle(makeSamples(subject));
+    }
+
+    public final MetricFamilySamples sampleAll(Iterable<T> subjects) {
+        List<Sample> samples;
+        if (subjects instanceof Collection) {
+            samples = new ArrayList<>(((Collection<T>) subjects).size());
+        } else {
+            samples = new ArrayList<>();
+        }
+
+        for (T subject : subjects) {
+            samples.addAll(makeSamples(subject));
+        }
+
+        return bundle(samples);
+    }
+
+    public static <T> Builder<T> build(String name, ToDoubleFunction<T> f) {
+        return new Builder<>("", "", (unit, help) -> new SimplePullGauge<>(name, unit, help, f));
+    }
+
+    public static <T> Builder<T> build(String name, LabeledPullGauge.Sampler<T> f) {
+        return new Builder<>("", "", (unit, help) -> new LabeledPullGauge<>(name, unit, help, v -> singletonList(f.apply(v))));
+    }
+
+    public static <T> Builder<T> build(String name, LabeledPullGauge.MultiSampler<T> f) {
+        return new Builder<>("", "", (unit, help) -> new LabeledPullGauge<>(name, unit, help, f));
+    }
+
+    public static final class Builder<T> {
+        private final String unit;
+        private final String help;
+        private final BiFunction<String, String, PullGauge<T>> ctor;
+
+        private Builder(String unit, String help, BiFunction<String, String, PullGauge<T>> ctor) {
+            this.unit = unit;
+            this.help = help;
+            this.ctor = ctor;
+        }
+
+        public Builder<T> unit(String unit) {
+            return new Builder<>(unit, help, ctor);
```

Comment:
Should not be returning a new instance here - field should be updated and `this` should be returned.

Reply:
It's an immutable builder

Reply:
We don't use immutable builders - that defeats the usual purpose of a builder.

Reply:
I start to remember why I don't usually touch java any more :D Anyway: Question now becomes: Do I align the builder with other collector builders from prometheus or do I align it with builders from sponge?